src/class/basic.rs

```diff
@@ -502,27 +487,17 @@ where
         where
             T: for<'p> PyObjectRichcmpProtocol<'p>,
         {
-            let py = Python::assume_gil_acquired();
-            let _pool = crate::GILPool::new(py);
-            let slf = py.from_borrowed_ptr::<crate::PyCell<T>>(slf);
-            let arg = py.from_borrowed_ptr::<PyAny>(arg);
-
-            match slf.try_borrow() {
-                Ok(borrowed_slf) => {
-                    let res = match extract_op(op) {
-                        Ok(op) => match arg.extract() {
-                            Ok(arg) => borrowed_slf.__richcmp__(arg, op).into(),
-                            Err(e) => Err(e),
-                        },
-                        Err(e) => Err(e),
-                    };
-                    match res {
-                        Ok(val) => val.into_py(py).into_ptr(),
-                        Err(e) => e.restore_and_null(py),
-                    }
-                }
-                Err(e) => PyErr::from(e).restore_and_null(py),
-            }
+            let py = crate::Python::assume_gil_acquired();
```

Comment:
I might have missed this in the thread, but is `assume_gil_acquired` sound without calling `GILPool::new` beforehand? If so, could we document the reason somewhere?

Reply:
In all current code, `GILPool::new` is called immediately after, like this:

```rust
let py = Python::assume_gil_acquired();
let pool = GILPool::new(py);
...
```

What I'm changing it to in this PR is now

```rust
let py = Python::assume_gil_acquired();
catch_unwind!(py, {
    let pool = GILPool::new(py);
    ...
})
...
```

However you've rightly noticed I have accidentally lost the `GILPool::new` here. I will fix this and also carefully check the rest of the changes I made.

I can't comment on whether this usage of `assume_gil_acquired` is sound, but I have other soundness concerns with it. See #800